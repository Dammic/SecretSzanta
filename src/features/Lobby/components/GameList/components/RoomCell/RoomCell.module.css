@import 'variables/variables.css';

.room {
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 100%;
    border: 1px solid var(--game-list-element-border);
    padding-left: 5px;
    padding-right: 5px;
    color: var(--game-list-element-font);
}

.rightSide {
    display: flex;
    flex-direction: column;
    align-items: center;
    height: 100%;
}

.playerCount {
    font-size: 14px;
    position: relative;
    top: 2px;
}

.button {
    flex-grow: 1;
}

.icon {
    margin-left: 4px;
    font-size: 11px;
}

.roomName {
    overflow: hidden;
    text-overflow: ellipsis;
    padding-right: 20px;
}

@keyframes appearFromRight {
    from { position: relative; right: -10%; opacity: 0; }
    to { position: relative; right: 0px; opacity: 1.0; }
}

.passwordInput {
    animation-name: appearFromRight;
    animation-duration: 1s;
}
