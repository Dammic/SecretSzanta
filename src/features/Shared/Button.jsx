import React from 'react'
import PropTypes from 'prop-types'

export const Button = ({ label, onClick }) => {
    return (
        <div
            className="fabulous-button lined thick"
            onClick={onClick}
            role="button"
            tabIndex="0"
        >
            {label}
        </div>
    )
}

Button.propTypes = {
    label: PropTypes.string,
    onClick: PropTypes.func,
}
