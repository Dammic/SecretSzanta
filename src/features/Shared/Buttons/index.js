export { default as Button } from './Button/Button'
export { default as PrimaryButton } from './PrimaryButton/PrimaryButton'
export { default as FancyButton } from './FancyButton/FancyButton'
