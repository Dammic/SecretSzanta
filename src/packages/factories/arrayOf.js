import faker from 'faker'
import { times } from 'lodash'

export const arrayOf = (callback, amount, options = { seed: null }) => {
    if (options.seed) {
        faker.seed(options.seed)
    }

    const result = []
    times(amount, () => result.push(callback()))
    return result
}
