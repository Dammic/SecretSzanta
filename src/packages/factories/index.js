export { arrayOf } from './arrayOf'
export { aMessage } from './messages'
export { aRoom, aRoomsList, aPlayer, aPlayersList } from './lobby'
