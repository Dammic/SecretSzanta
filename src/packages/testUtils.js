import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

const shallowRenderer = new ShallowRenderer()

export const expectMatchingSnapshot = (element) => {
    const renderedElement = renderer.create(element)
    expect(renderedElement).toMatchSnapshot()
    renderedElement.unmount()
}

export const expectShallowMatchingSnapshot = (element) => {
    const renderedElement = shallowRenderer.render(element)
    expect(renderedElement).toMatchSnapshot()
    renderedElement.unmount()
}
