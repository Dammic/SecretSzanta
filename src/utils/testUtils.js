import { mount } from 'enzyme'

export const expectMatchingSnapshot = (element) => {
    const renderedComponent = mount(element)
    expect(renderedComponent).toMatchSnapshot()
}
