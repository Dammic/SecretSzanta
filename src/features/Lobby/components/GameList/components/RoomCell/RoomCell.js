import React, { useCallback, useRef, useState } from 'react'
import { Icon } from '../../../../../Shared/Icon'
import { ControlButton } from '../../../../../Shared/Buttons'

import styles from './RoomCell.css'

export const RoomCell = ({ room: { roomName, playersCount, roomId, maxPlayers, hasPassword }, onJoin }) => {
    const [showPasswordInput, setPasswordInputVisibility] = useState(false)
    const [password, setPassword] = useState('')

    const onClick = useCallback(
        () => {
            if (hasPassword && !showPasswordInput) {
                setPasswordInputVisibility(true)
                return
            }

            onJoin(roomId, password)
        },
        [hasPassword, setPasswordInputVisibility, roomId, password],
    )

    return (
        <div className={styles.room}>
            <div className={styles.roomName}>{roomName}</div>
            {showPasswordInput && (
                <input
                    autoFocus={true}
                    className={styles.passwordInput}
                    onChange={event => setPassword(event.target.value)}
                    onKeyDown={event => event.key === 'Enter' && onJoin(roomId, password)}
                    placeholder="Type password to room"
                    type="password"
                    value={password}
                />
            )}
            <div className={styles.rightSide}>
                <ControlButton onClick={onClick} className={styles.button}>
                    join
                    {hasPassword && <Icon name="fa-lock" className={styles.icon} />}
                </ControlButton>
                <span className={styles.playerCount}>
                    {playersCount}/{maxPlayers}
                    <Icon name="fa-user" className={styles.icon} />
                </span>
            </div>
        </div>
    )
}
