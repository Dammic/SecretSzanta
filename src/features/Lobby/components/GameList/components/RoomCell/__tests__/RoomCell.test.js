import { aRoom } from 'packages/factories'
import { expectMatchingSnapshot } from 'packages/testUtils'
import React from 'react'
import { mount } from 'enzyme'
import { ControlButton } from '../../../../../../Shared/Buttons'

import { RoomCell } from '../RoomCell'

const setupProps = (propsOverrides = {}, renderMethod = mount) => {
    const props = {
        room: aRoom(),
        onJoin: jest.fn(),
        ...propsOverrides,
    }
    const component = renderMethod(<RoomCell {...props} />)
    return { props, component }
}

describe('<RoomCell />', () => {
    it('displays password input after attept to join secured room', () => {
        const { props, component } = setupProps({ room: aRoom({ hasPassword: true }) })

        component.find(ControlButton).first().simulate('click')

        component.update()
        expect(component.find('input').exists()).toEqual(true)
    })

    describe('calls onJoin callback when room', () => {
        it('has no password', () => {
            const { props: { onJoin }, component } = setupProps()

            component.find(ControlButton).first().simulate('click')

            expect(onJoin).toHaveBeenCalled()
        })

        it('has password, it is called with password passed', () => {
            const testRoom = aRoom({ hasPassword: true });
            const password = 'somePassword';

            const { props: { onJoin }, component } = setupProps({ room: testRoom })

            component.find(ControlButton).first().simulate('click')

            component.update()
            component.find('input').simulate('change', { target: { value: password } })

            component.find(ControlButton).first().simulate('click')
            expect(onJoin).toHaveBeenCalledWith(testRoom.roomId, password)
        })
    })

    it('should match snippet when romm has password', () => {
        const { props } = setupProps({ room: aRoom({ hasPassword: true }, { seed: 1 }) })

        expectMatchingSnapshot(<RoomCell {...props} />)
    })

    it('should match snippet when romm doesn\'t have password', () => {
        const { props } = setupProps({ room: aRoom({ hasPassword: false }, { seed: 1 }) })

        expectMatchingSnapshot(<RoomCell {...props} />)
    })
})
