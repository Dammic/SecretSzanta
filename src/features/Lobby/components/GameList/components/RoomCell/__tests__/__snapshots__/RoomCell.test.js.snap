// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RoomCell /> should match snippet when romm doesn't have password 1`] = `
<div
  className="room"
>
  <div
    className="roomName"
  >
    Corporate
  </div>
  <div
    className="rightSide"
  >
    <div
      className="btn control button"
      onClick={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="buttonText"
      >
        join
      </span>
    </div>
    <span
      className="playerCount"
    >
      /
      <i
        className="fa fa-user icon"
      />
    </span>
  </div>
</div>
`;

exports[`<RoomCell /> should match snippet when romm has password 1`] = `
<div
  className="room"
>
  <div
    className="roomName"
  >
    Corporate
  </div>
  <div
    className="rightSide"
  >
    <div
      className="btn control button"
      onClick={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="buttonText"
      >
        join
        <i
          className="fa fa-lock icon"
        />
      </span>
    </div>
    <span
      className="playerCount"
    >
      /
      <i
        className="fa fa-user icon"
      />
    </span>
  </div>
</div>
`;
